cmake/modules/dts.cmake

```diff
@@ -69,246 +141,362 @@ foreach(dts_dir ${DTS_ROOT})
-  toolchain_parse_make_rule(${DTS_DEPS}
-    include_files # Output parameter
-    )
+  # If we have a system devicetree, the user must specify the path to the
+  # execution domain that they want to select via SYSDT_EXECUTION_DOMAIN.
+  if(NOT DEFINED SYSDT_EXECUTION_DOMAIN)
+    message(FATAL_ERROR "Found system devicetree file ${BOARD_SYSDTS}, but SYSDT_EXECUTION_DOMAIN was not defined. Set this variable to the path to the execution domain you wish to use.")
+  endif()
 
-  set_property(DIRECTORY APPEND PROPERTY
-    CMAKE_CONFIGURE_DEPENDS
-    ${include_files}
-    ${GEN_DEFINES_SCRIPT}
-    ${GEN_DRIVER_KCONFIG_SCRIPT}
-    ${DTS_CMAKE_SCRIPT}
+  execute_process(COMMAND ${PYTHON_EXECUTABLE} ${GEN_DTSI_FROM_SYSDTS_SCRIPT}
+    --sysdts "${SYSDT_POST_CPP}"
+    --domain "${SYSDT_EXECUTION_DOMAIN}"
+    --sysdt-pickle-out "${SYSDT_PICKLE}"
+    --dts-out "${SYSDT_OUTPUT_DTS}"
+    RESULT_VARIABLE ret
     )
+  if(NOT ret EQUAL 0)
+    message(FATAL_ERROR "Failed to generate .dtsi from system devicetree: ${gen_dt_from_sysdt_result}")
+  endif()
+endif(BOARD_HAS_SYSDTS)
 
+#
+# Halt execution early if there is no devicetree or system devicetree.
+#
+
+if(BOARD_HAS_SYSDTS)
+  # When a system devicetree is present, we'll use the output devicetree
+  # in the build directory as DTS_SOURCE.
   #
-  # Run gen_defines.py to create a header file, zephyr.dts, and edt.pickle.
-  #
+  # TODO: how can we support board revisions in system devicetree?
 
-  string(REPLACE ";" " " EXTRA_DTC_FLAGS_RAW "${EXTRA_DTC_FLAGS}")
-  set(CMD_EXTRACT ${PYTHON_EXECUTABLE} ${GEN_DEFINES_SCRIPT}
-  --dts ${DTS_POST_CPP}
-  --dtc-flags '${EXTRA_DTC_FLAGS_RAW}'
-  --bindings-dirs ${DTS_ROOT_BINDINGS}
-  --header-out ${DEVICETREE_GENERATED_H}.new
-  --dts-out ${ZEPHYR_DTS}.new # for debugging and dtc
-  --edt-pickle-out ${EDT_PICKLE}
-  ${EXTRA_GEN_DEFINES_ARGS}
-  )
+  if(DEFINED DTS_SOURCE)
+    message(FATAL_ERROR "do not define DTS_SOURCE; we have a system devicetree")
+  endif()
 
-  execute_process(
-    COMMAND ${CMD_EXTRACT}
-    WORKING_DIRECTORY ${PROJECT_BINARY_DIR}
-    RESULT_VARIABLE ret
-    )
-  if(NOT "${ret}" STREQUAL "0")
-    message(STATUS "In: ${PROJECT_BINARY_DIR}, command: ${CMD_EXTRACT}")
-    message(FATAL_ERROR "gen_defines.py failed with return code: ${ret}")
+  set(DTS_SOURCE ${SYSDT_OUTPUT_DTS})
+
+  # HACK
+  set(domain_overlay ${BOARD_DIR}${SYSDT_EXECUTION_DOMAIN}.overlay)
+  if(EXISTS ${domain_overlay})
+    list(APPEND DTS_SOURCE ${domain_overlay})
+  endif()
+else()
+  # TODO: What to do about non-posix platforms where NOT CONFIG_HAS_DTS (xtensa)?
+  # Drop support for NOT CONFIG_HAS_DTS perhaps?
+
+  set_ifndef(DTS_SOURCE ${BOARD_DIR}/${BOARD}.dts)
+  if(EXISTS ${DTS_SOURCE})
+    # We found a devicetree. Check for a board revision overlay.
+    if(BOARD_REVISION AND EXISTS ${BOARD_DIR}/${BOARD}_${BOARD_REVISION_STRING}.overlay)
+      list(APPEND DTS_SOURCE ${BOARD_DIR}/${BOARD}_${BOARD_REVISION_STRING}.overlay)
```

Comment:
(From above) similar to this, boards will either be v2 systemdt or v1 non-systemdt so should be able to have v2 specific handling